libsolidity/ast/TypeProvider.cpp

```diff
@@ -0,0 +1,355 @@
+{
+	if (_type->location() == _location && _type->isPointer() == _isPointer)
+		return _type;
+
+	// TODO: re-use existing objects if present (requires deep equality check)
+
+	m_referenceTypes.emplace_back(_type->copyForLocation(_location, _isPointer));
+	return m_referenceTypes.back().get();
+}
+
+FunctionType const* TypeProvider::functionType(FunctionDefinition const& _function, bool _isInternal)
+{
+	// TODO: reuse existing types meaningful?
+
+	m_functionTypes.emplace_back(make_unique<FunctionType>(_function, _isInternal));
+	return m_functionTypes.back().get();
+}
+
+FunctionType const* TypeProvider::functionType(VariableDeclaration const& _varDecl)
+{
+	// TODO: reuse existing types meaningful?
+
+	m_functionTypes.emplace_back(make_unique<FunctionType>(_varDecl));
+	return m_functionTypes.back().get();
+}
+
+FunctionType const* TypeProvider::functionType(EventDefinition const& _def)
+{
+	// TODO: reuse existing types meaningful?
+
+	m_functionTypes.emplace_back(make_unique<FunctionType>(_def));
+	return m_functionTypes.back().get();
+}
+
+FunctionType const* TypeProvider::functionType(FunctionTypeName const& _typeName)
+{
+	// TODO: reuse existing types meaningful?
+
+	m_functionTypes.emplace_back(make_unique<FunctionType>(_typeName));
+	return m_functionTypes.back().get();
+}
+
+FunctionType const* TypeProvider::functionType(
+	strings const& _parameterTypes,
+	strings const& _returnParameterTypes,
+	FunctionType::Kind _kind,
+	bool _arbitraryParameters,
+	StateMutability _stateMutability
+)
+{
+	// TODO: reuse existing types meaningful?
+
+	m_functionTypes.emplace_back(make_unique<FunctionType>(
+		_parameterTypes, _returnParameterTypes,
+		_kind, _arbitraryParameters, _stateMutability
+	));
+	return m_functionTypes.back().get();
+}
+
+FunctionType const* TypeProvider::functionType(
+	TypePointers const& _parameterTypes,
+	TypePointers const& _returnParameterTypes,
+	strings _parameterNames,
+	strings _returnParameterNames,
+	FunctionType::Kind _kind,
+	bool _arbitraryParameters,
+	StateMutability _stateMutability,
+	Declaration const* _declaration,
+	bool _gasSet,
+	bool _valueSet,
+	bool _bound
+)
+{
+	// TODO: reuse existing types meaningful?
+
+	m_functionTypes.emplace_back(make_unique<FunctionType>(
+		_parameterTypes,
+		_returnParameterTypes,
+		_parameterNames,
+		_returnParameterNames,
+		_kind,
+		_arbitraryParameters,
+		_stateMutability,
+		_declaration,
+		_gasSet,
+		_valueSet,
+		_bound
+	));
+	return m_functionTypes.back().get();
+}
+
+RationalNumberType const* TypeProvider::rationalNumberType(rational const& _value, Type const* _compatibleBytesType)
+{
+	// TODO: reuse existing types (We currently don't expose `rational m_value`.)
+
+	m_rationalNumberTypes.emplace_back(make_unique<RationalNumberType>(_value, _compatibleBytesType));
+	return m_rationalNumberTypes.back().get();
+}
+
+ArrayType const* TypeProvider::arrayType(DataLocation _location, bool _isString)
+{
+	// TODO: reuse existing types
```

Comment:
I think this will be really hard. Until we do that, it might be better to just have a single vector of `Type*` instead of different vectors of different types. Even then we can even implement `operator<` and first discriminate on category first.

Reply:
Also, create, add to set, return could be implemented as a template for all (easy to construct) types at a time.